python/sidecar/ingest.py

```diff
@@ -238,25 +237,16 @@ def create_references(self) -> List[Reference]:
         failures = self._create_references_for_grobid_failures()
         return references + failures
 
-    def create_response_from_references(self, references: List[Reference]) -> dict:
+    def create_response_from_references(self, references: List[Reference]) -> IngestResponse:
         """
         Creates a Response object from a list of Reference objects
         :param references: List[Reference]
         :return: Response
         """
-        prepared_references = []
-        for ref in references:
-            # no need to include reference abstract, contents, or chunks in response
```

Comment:
@danvk it seems this doesn't work for nested dataclasses. for example:
```python
    references = [
        Reference(
            source_filename="some_file.pdf",
            filename_md5="abc123",
            title="Some title",
            authors=[
                Author(full_name="John Doe"),
                Author(full_name="Jane Doe"),
            ],
            chunks=[
                Chunk(text="This is some text."),
                Chunk(text="This is some more text."),
            ]
        ),
        Reference(
            source_filename="another_file.pdf",
            filename_md5="abcdefg123",
            title="Another title",
            authors=[
                Author(full_name="John Doe"),
                Author(full_name="Jane Doe"),
            ],
            chunks=[
                Chunk(text="Some text about something interesting."),
                Chunk(text="More text about something."),
            ]
        )
    ]
```

```python
In [26]: ref = references[0]

In [27]: ref.to_json()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[27], line 1
----> 1 ref.to_json()

File ~/Library/Caches/pypoetry/virtualenvs/refstudio-iTeERiwC-py3.10/lib/python3.10/site-packages/dataclasses_jsonschema/__init__.py:1017, in JsonSchemaMixin.to_json(self, omit_none, validate, **json_kwargs)
   1016 def to_json(self, omit_none: bool = True, validate: bool = False, **json_kwargs) -> str:
-> 1017     return json.dumps(self.to_dict(omit_none, validate), **json_kwargs)

File ~/.pyenv/versions/3.10.11/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/__init__.py:231, in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    226 # cached encoder
    227 if (not skipkeys and ensure_ascii and
    228     check_circular and allow_nan and
    229     cls is None and indent is None and separators is None and
    230     default is None and not sort_keys and not kw):
--> 231     return _default_encoder.encode(obj)
    232 if cls is None:
    233     cls = JSONEncoder

File ~/.pyenv/versions/3.10.11/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/encoder.py:199, in JSONEncoder.encode(self, o)
    195         return encode_basestring(o)
    196 # This doesn't pass the iterator directly to ''.join() because the
    197 # exceptions aren't as detailed.  The list call should be roughly
    198 # equivalent to the PySequence_Fast that ''.join() would do.
--> 199 chunks = self.iterencode(o, _one_shot=True)
    200 if not isinstance(chunks, (list, tuple)):
    201     chunks = list(chunks)

File ~/.pyenv/versions/3.10.11/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/encoder.py:257, in JSONEncoder.iterencode(self, o, _one_shot)
    252 else:
    253     _iterencode = _make_iterencode(
    254         markers, self.default, _encoder, self.indent, floatstr,
    255         self.key_separator, self.item_separator, self.sort_keys,
    256         self.skipkeys, _one_shot)
--> 257 return _iterencode(o, 0)

File ~/.pyenv/versions/3.10.11/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/encoder.py:179, in JSONEncoder.default(self, o)
    160 def default(self, o):
    161     """Implement this method in a subclass such that it returns
    162     a serializable object for ``o``, or calls the base implementation
    163     (to raise a ``TypeError``).
   (...)
    177 
    178     """
--> 179     raise TypeError(f'Object of type {o.__class__.__name__} '
    180                     f'is not JSON serializable')

TypeError: Object of type Author is not JSON serializable
```

Calling `to_dict()` will run, but `Author` will not be serialized and will remain as a dataclass. maybe i'm doing something wrong?

i know that `asdict` will serialize everything.

Reply:
i think this was ultimately related to using `str | None` vs `Optional[str]`